src/base.jl

```diff
@@ -48,20 +46,68 @@ size_nibbles(::Type{T}) where T = ceil(log2(sizeof(T))/4)
 
 size_mask(T) = UInt(exp2(4*size_nibbles(T)) - 1)
 
+@inline function Base.isascii(s::ShortString{T}) where T
+    val = s.size_content << (8*size_nibbles(T))
+    for i in 1:sizeof(T)
+        iszero(val & 0x80) || return false
+        val <<= 8  # first byte never matters as will always be
+    end
+    return true
+end
 
-# function Base.getindex(s::ShortString, i::Integer)
-#     getindex(String(s), i)
-# end
+function Base.length(s::ShortString{T}) where T
+    isascii(s) && return ncodeunits(s)
+
+    # else have to do it the hard way:
+    i = 0
+    len = 0
+    while i < ncodeunits(s)
+        shifted = s.size_content >> (8*(sizeof(T) - i))
+        i += if shifted % UInt8 <= 0x7f  # 1 byte character
+            1
+        elseif shifted % UInt16 <= 0x7ff  # 2 byte character
+            2
+        elseif shifted % UInt32 <= 0xffff  # 3 byte character
+            3
+        else  # 4 byte character
+            4
+        end
+        len += 1
+    end
+    return len
+end
 
-# function Base.getindex(s::ShortString, args...; kwargs...)
-#     getindex(String(s), args...; kwargs...)
-# end
+@inline function Base.iterate(s::ShortString{T}, i::Integer=1) where T
+    i > ncodeunits(s) && return nothing
+    shifted = s.size_content >> (8*(sizeof(T) - i))
+    if shifted % UInt8 <= 0x7f  # 1 byte character
+        return Char(shifted % UInt8), i+1
+    elseif shifted % UInt16 <= 0x7ff  # 2 byte character
+        return Char(shifted % UInt16 % UInt32), i+2
+    elseif shifted % UInt32 <= 0xffff  # 3 byte character
+        return Char(shifted % UInt32 & 0xffff), i+3
+    else  # 4 byte character
+        return Char(shifted % UInt32), i+4
+    end
+end
 
-Base.collect(s::ShortString) = collect(String(s))
+function ==(s::ShortString{S}, b::AbstractString) where S
+    ncodeunits(b) == ncodeunits(s) || return false
```

Comment:
This check is incorrect, the number of code units is dependent on how it is encoded.
I think it would be useful for ShortString to have another parameter, which is the type of string that is being encoded, such as `String`, or `ASCIIStr`, `UTF8Str`, `Latin1Str`, etc. (or even the old `LegacyString` types)
Maybe some traits could be used / defined so that these could be compared in the most efficient fashion.
(As an example, an Emoji takes 4 codeunits in String or UTF8Str, 2 codeunits with UTF16Str, and 1 codeunit with UTF32Str)

Reply:
true, this actually shouldn't be in this PR. it  got caught up from #16

Reply:
> number of code units is dependent on how it is encoded.

this is not something I understand. Is there anything I can read up?

Reply:
Some encoding schemes are byte oriented, such as UTF-8, but UCS2 and UTF-16 use 16-bit codeunits, and UTF-32 uses 32-bit codeunits.

Reply:
UTF-16 is actually a lot more efficient for storing non Western European languages than UTF-8.  Most all of the common Chinese, Korean, and Japanese characters only take 2 bytes, instead of 3 bytes in UTF-8, so you'd be able to store 7 characters in a UInt128, instead of just 5.

Reply:
The largest size in ShortStrings would store up to 63 UTF-16 characters (but only up to 42 of the same Asian characters using UTF-8)